src/include/plan_node/abstract_plan_node.h

```diff
@@ -0,0 +1,246 @@
+#pragma once
+
+#include <cstdint>
+#include <functional>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "common/hash_util.h"
+#include "common/json.h"
+#include "plan_node/output_schema.h"
+#include "plan_node/plan_node_defs.h"
+
+namespace terrier::plan_node {
+
+/**
+ * An abstract plan node should be the base class for (almost) all plan nodes
+ */
+class AbstractPlanNode {
+ protected:
+  /**
+   * Base builder class for plan nodes
+   * @tparam ConcreteType
+   */
+  template <class ConcreteType>
+  class Builder {
+   public:
+    virtual ~Builder() = default;
+
+    /**
+     * @param child child to be added
+     * @return builder object
+     */
+    ConcreteType &AddChild(std::unique_ptr<AbstractPlanNode> child) {
+      children_.emplace_back(std::move(child));
+      return *dynamic_cast<ConcreteType *>(this);
+    }
+
+    /**
+     * @param output_schema output schema for plan node
+     * @return builder object
+     */
+    ConcreteType &SetOutputSchema(std::shared_ptr<OutputSchema> output_schema) {
+      output_schema_ = output_schema;
+      return *dynamic_cast<ConcreteType *>(this);
+    }
+
+    /**
+     * @param cardinality estimated cardinality of output for the plan node
+     * @return builder object
+     */
+    ConcreteType &SetEstimatedCardinality(int cardinality) {
+      estimated_cardinality_ = cardinality;
+      return *dynamic_cast<ConcreteType *>(this);
+    }
+
+   protected:
+    /**
+     * child plans
+     */
+    std::vector<std::unique_ptr<AbstractPlanNode>> children_;
+    /**
+     * schema describing output of the node
+     */
+    std::shared_ptr<OutputSchema> output_schema_;
+    /**
+     * estimated cardinality of output for node
+     */
+    uint32_t estimated_cardinality_ = 0;
```

Comment:
I don't think we want this in here since it will change over time as the database changes. It should be stored in a context object.